README.md
ANOM
====

R code creating analysis-of-mean decision charts.


inst/NEWS
#########################
### ANOM package news ###
#########################

Please report weird or unexpected behavior to Philip Pallmann <p.pallmann@lancaster.ac.uk>

#########################

Version 0.5 (10 April 2017)

- Added more flexibility for tweaking font sizes, line widths, colours, and background grid lines.
- Updated CITATION file.

#########################

Version 0.4.3 (15 February 2016)

- Included a CITATION file.
- Updated references in the help files and vignette.

#########################

Version 0.4.2 (13 October 2015)

- Fixed a bug with printp.

#########################

Version 0.4.1 (17 August 2015)

- Fixed a minor bug with the background color in ANOMgen.

#########################

Version 0.4 (12 August 2015)

- Added an option to switch between white or gray background.
- Moved the PDF vignette from folder inst/doc to vignettes.
- Added lme4 and nlme to Suggests.
- Changed e-mail of the package maintainer.

#########################

Version 0.3.1 (11 February 2014)

- Minor changes to the references.

#########################

Version 0.3 (19 November 2013)

- Added a vignette including a couple of worked examples. Type 'vignette("ANOM")' to check it out!

R/ANOM.R
ANOM <- function(mc, xlabel=NULL, ylabel=NULL, printn=TRUE, printp=TRUE,
                 stdep=NULL, stind=NULL, pst=NULL, pbin=NULL, bg="white",
                 bgrid=TRUE, axlsize=18, axtsize=25, npsize=5, psize=5,
                 lwidth=1, dlstyle="dashed", fillcol="darkgray"){
  
  if(!(class(mc)[1] %in% c("glht", "SimCi", "mctp", "binomRDci"))){
    stop("Please insert an object of class 'glht', 'SimCi', 'mctp', or 'binomRDci' for mc!")
  }
  
  if(class(mc)[1]=="glht"){
    
    if(mc$type!="GrandMean"){
      stop("For ANOM you need a 'GrandMean' contrast matrix!")
    }
    
    modclass <- attr(mc$model, "class")[1]
    
    if(modclass=="aov" | modclass=="lm" | modclass=="glm"){
      dd <- attr(attr(mc$model$terms, "factors"), "dimnames")[[1]][1]
      ii <- attr(attr(mc$model$terms, "factors"), "dimnames")[[1]][2]
      dep <- mc$model$model[dd][, 1]
      ind <- mc$model$model[ii][, 1]    
    }
    
    if(modclass=="lme"){
      dd <- attr(attr(mc$model$terms, "factors"), "dimnames")[[1]][1]
      ii <- attr(attr(mc$model$terms, "factors"), "dimnames")[[1]][2]
      dep <- mc$model$data[, dd]
      ind <- mc$model$data[, ii]
    }
    
    if(modclass=="lmerMod"){
      dd <- rownames(attr(attr(mc$model@frame, "terms"), "factors"))[1]
      ii <- rownames(attr(attr(mc$model@frame, "terms"), "factors"))[2]
      dep <- mc$model@frame[, dd]
      ind <- mc$model@frame[, ii]
    }
    
    ss <- as.vector(tapply(ind, ind, length))
    ci <- confint(mc)
    
    if(modclass!="glm"){
      means <- as.vector(tapply(dep, ind, mean))
      mea <- mean(dep)
    }
    
    if(modclass=="glm"){
      if(mc$model$family$family=="poisson"){
        cit <- apply(ci$confint, 2, poisson()$linkinv)
        m <- mean(poisson()$linkinv(predict(mc$model)))
      }
      if(mc$model$family$family=="binomial"){
        cit <- apply(ci$confint, 2, binomial()$linkinv)
        m <- mean(binomial()$linkinv(predict(mc$model)))
      }
    }
    
    if(is.null(xlabel)==TRUE){
      xlabel <- ii
    }else{
      xlabel <- xlabel
    }
    
    if(is.null(ylabel)==TRUE){
      ylabel <- dd
    }else{
      ylabel <- ylabel
    }
    
    if(printp==TRUE){
      pvals <- summary(mc)$test$pvalues[1:length(ss)]
    }else{
      pvals <- NULL
    }
    
    bg <- match.arg(bg, choices=c("gray", "grey", "white"))
    
    if(modclass!="glm"){
      if(modclass=="lmerMod"){
        ANOMgen(mu=means, n=ss, lo=ci$confint[, "lwr"], up=ci$confint[, "upr"],
                names=levels(mc$model@frame[, ii]), alternative=mc$alternative,
                xlabel=xlabel, ylabel=ylabel, printn=printn, p=pvals, bg=bg,
                bgrid=bgrid, axlsize=axlsize, axtsize=axtsize, npsize=npsize,
                psize=psize, lwidth=lwidth, dlstyle=dlstyle, fillcol=fillcol)
      }else{
        ANOMgen(mu=means, n=ss, lo=ci$confint[, "lwr"], up=ci$confint[, "upr"],
                names=mc$model$xlevels[[1]], alternative=mc$alternative,
                xlabel=xlabel, ylabel=ylabel, printn=printn, p=pvals, bg=bg,
                bgrid=bgrid, axlsize=axlsize, axtsize=axtsize, npsize=npsize,
                psize=psize, lwidth=lwidth, dlstyle=dlstyle, fillcol=fillcol)
      }
    }
    
    if(modclass=="glm"){
      
      if(mc$model$family$family=="poisson"){
        if(mc$alternative=="two.sided"){
          ANOMintern(mu=m*cit[, 1], n=ss, gm=m, lo=m*cit[, 2]-m, up=m*cit[, 3]-m,
                     names=mc$model$xlevels[[1]], alternative=mc$alternative,
                     xlabel=xlabel, ylabel=ylabel, printn=printn, p=pvals, bg=bg,
                     bgrid=bgrid, axlsize=axlsize, axtsize=axtsize, npsize=npsize,
                     psize=psize, lwidth=lwidth, dlstyle=dlstyle, fillcol=fillcol,
                     whichone="glm")
        }
        if(mc$alternative=="greater"){
          ANOMintern(mu=m*cit[, 1], n=ss, gm=m, lo=m*cit[, 2]-m, up=Inf,
                     names=mc$model$xlevels[[1]], alternative=mc$alternative,
                     xlabel=xlabel, ylabel=ylabel, printn=printn, p=pvals, bg=bg,
                     bgrid=bgrid, axlsize=axlsize, axtsize=axtsize, npsize=npsize,
                     psize=psize, lwidth=lwidth, dlstyle=dlstyle, fillcol=fillcol,
                     whichone="glm")
        }
        if(mc$alternative=="less"){
          ANOMintern(mu=m*cit[, 1], n=ss, gm=m, lo=-Inf, up=m*cit[, 3]-m,
                     names=mc$model$xlevels[[1]], alternative=mc$alternative,
                     xlabel=xlabel, ylabel=ylabel, printn=printn, p=pvals, bg=bg,
                     bgrid=bgrid, axlsize=axlsize, axtsize=axtsize, npsize=npsize,
                     psize=psize, lwidth=lwidth, dlstyle=dlstyle, fillcol=fillcol,
                     whichone="glm")
        }
      }
      if(mc$model$family$family=="binomial"){
        if(mc$alternative=="two.sided"){
          ANOMintern(mu=cit[, 1], n=ss, gm=m, lo=cit[ ,2]-m, up=cit[ ,3]-m,
                     names=mc$model$xlevels[[1]], alternative=mc$alternative,
                     xlabel=xlabel, ylabel=ylabel, printn=printn, p=pvals, bg=bg,
                     bgrid=bgrid, axlsize=axlsize, axtsize=axtsize, npsize=npsize,
                     psize=psize, lwidth=lwidth, dlstyle=dlstyle, fillcol=fillcol,
                     whichone="glm")
        }
        if(mc$alternative=="less"){
          ANOMintern(mu=cit[, 1], n=ss, gm=m, lo=-Inf, up=cit[ ,3]-m,
                     names=mc$model$xlevels[[1]], alternative=mc$alternative,
                     xlabel=xlabel, ylabel=ylabel, printn=printn, p=pvals, bg=bg,
                     bgrid=bgrid, axlsize=axlsize, axtsize=axtsize, npsize=npsize,
                     psize=psize, lwidth=lwidth, dlstyle=dlstyle, fillcol=fillcol,
                     whichone="glm")
        }
        if(mc$alternative=="greater"){
          ANOMintern(mu=cit[, 1], n=ss, gm=m, lo=cit[ ,2]-m, up=Inf,
                     names=mc$model$xlevels[[1]], alternative=mc$alternative,
                     xlabel=xlabel, ylabel=ylabel, printn=printn, p=pvals, bg=bg,
                     bgrid=bgrid, axlsize=axlsize, axtsize=axtsize, npsize=npsize,
                     psize=psize, lwidth=lwidth, dlstyle=dlstyle, fillcol=fillcol,
                     whichone="glm")
        }
      }
      
    }
    
  }
  
  if(class(mc)[1]=="SimCi"){
    
    if(mc$type!="GrandMean"){
      stop("For ANOM you need a 'GrandMean' contrast matrix!")
    }
    
    if(is.null(stdep)==TRUE){
      stop("Please insert a vector giving the values\n
           of the dependent variable for stdep!")
    }
    
    if(is.null(stind)==TRUE){
      stop("Please insert a numeric vector giving the values\n
           of the independent variable for stind!")
    }
    
    if(is.numeric(stdep)==FALSE){stop("The dependent variable must be numeric.")}
    
    if((length(stdep)==length(stind))==FALSE){
      stop("Dependent and independent variable must be vectors of equal length.")
    }
    
    stind <- as.factor(stind)
    ss <- as.vector(tapply(stdep, stind, length))
    
    if(mc$test.class=="ratios"){
      grame <- 100
    }else{
      grame <- mean(stdep)
    }
    
    if(is.null(xlabel)==TRUE){
      xlabel <- "group"
    }else{
      xlabel <- xlabel
    }
    
    if(is.null(ylabel)==TRUE){
      ylabel <- mc$resp
    }else{
      ylabel <- ylabel
    }
    
    if(printp==TRUE){
      
      if(class(pst)!="SimTest"){
        stop("Please insert an object of class 'SimTest' for pst,\n
             or set printp to 'FALSE'.")
      }
      
      ppp <- pst$p.val.adj[1:length(ss)]
      
      }else{
        
        ppp <- NULL
        
      }
    
    if(mc$test.class=="ratios"){
      ANOMintern(mu=100*as.vector(mc$estimate), n=ss, lo=100*as.vector(mc$lower),
                 up=100*as.vector(mc$upper), alternative=mc$alternative,
                 xlabel=xlabel, ylabel=ylabel, printn=printn, p=ppp, bg=bg,
                 bgrid=bgrid, axlsize=axlsize, axtsize=axtsize, npsize=npsize,
                 psize=psize, lwidth=lwidth, dlstyle=dlstyle, fillcol=fillcol,
                 whichone="ratio")
    }else{
      ANOMgen(mu=as.vector(mc$estimate)+grame, n=ss, lo=as.vector(mc$lower),
              up=as.vector(mc$upper), alternative=mc$alternative,
              xlabel=xlabel, ylabel=ylabel, printn=printn, p=ppp, bg=bg,
              bgrid=bgrid, axlsize=axlsize, axtsize=axtsize, npsize=npsize,
              psize=psize, lwidth=lwidth, dlstyle=dlstyle, fillcol=fillcol)
    }
    
    }
  
  if(class(mc)[1]=="mctp"){
    
    if(mc$input$type!="UserDefined"){
      stop("For ANOM you need a 'UserDefined' grand-mean-type contrast matrix!")
    }
    
    if(is.null(mc$Correlation)==TRUE){
      stop("Set the argument 'correlation' in function 'mctp()' to 'TRUE'.")
    }
    
    if(is.null(xlabel)==TRUE){
      xlabel <- as.character(mc$input$formula[3])
    }else{
      xlabel <- xlabel
    }
    
    if(is.null(ylabel)==TRUE){
      ylabel <- as.character(mc$input$formula[2])
    }else{
      ylabel <- ylabel
    }
    
    if(printp==TRUE){
      pvals <- mc$Analysis$p.Value
    }
    
    if(mc$input$alternative=="two.sided"){
      ANOMgen(mu=mc$Data.Info$Effect, n=mc$Data.Info$Size,
              lo=mc$Analysis$Lower, #abs(mc$Analysis$Estimator - mc$Analysis$Lower),
              up=mc$Analysis$Upper, #abs(mc$Analysis$Estimator - mc$Analysis$Upper),
              names=colnames(mc$Contrast), alternative=mc$input$alternative,
              xlabel=xlabel, ylabel=ylabel,
              printn=printn, p=pvals, bg=bg, bgrid=bgrid,
              axlsize=axlsize, axtsize=axtsize, npsize=npsize, psize=psize, lwidth=lwidth,
              dlstyle=dlstyle, fillcol=fillcol)
    }
    
    if(mc$input$alternative=="greater"){
      ANOMgen(mu=mc$Data.Info$Effect, n=mc$Data.Info$Size,
              lo=mc$Analysis$Lower, #abs(mc$Analysis$Estimator - mc$Analysis$Lower),
              up=Inf,
              names=colnames(mc$Contrast), alternative=mc$input$alternative,
              xlabel=xlabel, ylabel=ylabel,
              printn=printn, p=pvals, bg=bg, bgrid=bgrid,
              axlsize=axlsize, axtsize=axtsize, npsize=npsize, psize=psize, lwidth=lwidth,
              dlstyle=dlstyle, fillcol=fillcol)
    }
    
    if(mc$input$alternative=="less"){
      ANOMgen(mu=mc$Data.Info$Effect, n=mc$Data.Info$Size,
              lo=Inf,
              up=mc$Analysis$Upper, #abs(mc$Analysis$Estimator - mc$Analysis$Upper),
              names=colnames(mc$Contrast), alternative=mc$input$alternative,
              xlabel=xlabel, ylabel=ylabel,
              printn=printn, p=pvals, bg=bg, bgrid=bgrid,
              axlsize=axlsize, axtsize=axtsize, npsize=npsize, psize=psize, lwidth=lwidth,
              dlstyle=dlstyle, fillcol=fillcol)
    }
    
  }
  
  if(class(mc)[1]=="binomRDci"){
    
    if(attr(mc$cmat, "type")!="GrandMean"){
      stop("For ANOM you need a 'GrandMean' contrast matrix!")
    }
    
    if(is.null(xlabel)==TRUE){
      xlabel <- "group"
    }else{
      xlabel <- xlabel
    }
    
    if(is.null(ylabel)==TRUE){
      ylabel <- "probability of success"
    }else{
      ylabel <- ylabel
    }
    
    if(printp==TRUE){
      
      if(class(pbin)!="binomRDtest"){
        stop("Please insert an object of class 'binomRDtest' for pbin,\n
             or set printp to 'FALSE'.")
      }
      
      ppp <- pbin$p.val.adj
      
      }else{
        
        ppp <- NULL
        
      }
    
    ANOMgen(mu=mc$p, n=mc$n, lo=mc$conf.int[, "lower"],
            up=mc$conf.int[, "upper"], names=mc$names,
            alternative=mc$alternative, xlabel=xlabel, ylabel=ylabel,
            printn=printn, p=ppp, bg=bg, bgrid=bgrid,
            axlsize=axlsize, axtsize=axtsize, npsize=npsize, psize=psize, lwidth=lwidth,
            dlstyle=dlstyle, fillcol=fillcol)
    
  }
  
}

R/ANOMgen.R
ANOMgen <- function(mu, n=NULL, gm=NULL, lo, up, names=NULL, alternative="two.sided",
                    xlabel="Group", ylabel="Endpoint", printn=TRUE, p=NULL, bg="white",
                    bgrid=TRUE, axlsize=18, axtsize=25, npsize=5, psize=5, lwidth=1,
                    dlstyle="dashed", fillcol="darkgray"){
  
  if(is.null(gm)){
    gm <- weighted.mean(mu, n)
  }
  
  if(!(is.null(n)) & !(is.null(gm))){
    check <- abs(weighted.mean(mu, n) - gm)
    if(check > 0.01){
      stop("The 'gm' value you inserted is not the grand mean\n
           computed from your 'mu' and 'n' values!")
    }
    }
  
  grp <- 1:length(mu)
  grpf <- as.factor(grp)
  
  if(!(is.null(names))){
    names <- as.factor(names)
  }else{
    names <- grpf
  }
  
  if(is.null(n)){
    printn <- FALSE
  }
  
  if(!(is.null(p))){
    printp <- TRUE
    p <- paste("p=", round(p, 3), sep="")
    p[p=="p=0"] <- "p<0.001"
  }else{
    printp <- FALSE
  }
  
  dir <- match.arg(alternative, choices=c("two.sided", "greater", "less"))
  
  if(dir=="two.sided"){
    if(all(lo==-Inf) | all(lo==-1) | all(lo==0) | all(up==1) | all(up==Inf)){
      warning("Is your alternative really 'two.sided'? Doesn't seem so!")
    }
  }
  
  ldl <- gm - abs((mu - gm) - up)
  udl <- gm + abs((mu - gm) - lo)
  
  if(printp==TRUE){
    if(printn==TRUE){
      set <- data.frame(mu, gm, lo, up, ldl, udl, grp, grpf, names, n, p)
    }else{
      set <- data.frame(mu, gm, lo, up, ldl, udl, grp, grpf, names, p)
    }
  }else{
    if(printn==TRUE){
      set <- data.frame(mu, gm, lo, up, ldl, udl, grp, grpf, names, n)
    }else{
      set <- data.frame(mu, gm, lo, up, ldl, udl, grp, grpf, names)
    }
  }
  
  bg <- match.arg(bg, choices=c("gray", "grey", "white"))
  
  if(bg=="white"){
    back <- theme_bw()
  }else{
    back <- theme_gray()
  }
  
  if(bgrid==TRUE){
    bgr <- element_line()
  }else{
    bgr <- element_blank()
  }
  
  if(dir=="two.sided"){
    
    basic <- ggplot(set, aes(x=grpf, y=mu)) +
      scale_x_discrete(labels=names) +
      xlab(xlabel) +
      ylab(ylabel) +
      geom_rect(aes(xmin=grp-0.5, xmax=grp+0.5, ymin=ldl, ymax=udl),
                alpha=0.5, fill=fillcol, linetype=0) +
      geom_segment(aes(x=grp-0.5, xend=grp+0.5, y=udl, yend=udl), size=lwidth, linetype=dlstyle) +
      geom_segment(aes(x=grp-0.5, xend=grp+0.5, y=ldl, yend=ldl), size=lwidth, linetype=dlstyle) +
      geom_segment(aes(x=0.5, xend=max(grp)+0.5, y=gm, yend=gm), size=lwidth) +
      geom_segment(aes(x=grp, xend=grp, y=mu, yend=gm), size=lwidth) +
      geom_point(size=psize) +
      annotate("text", label="LDL", x=max(grp)+0.4, y=ldl[max(grp)], size=4, vjust=1.5) +
      annotate("text", label="UDL", x=max(grp)+0.4, y=udl[max(grp)], size=4, vjust=-0.75) +
      ylim(min(min(mu), min(ldl))-(gm-min(min(mu), min(ldl)))/5,
           max(max(mu), max(udl))+(max(max(mu), max(udl))-gm)/5) +
      back +
      theme(axis.text.x=element_text(size=axlsize),
            axis.text.y=element_text(size=axlsize),
            axis.title.x=element_text(size=axtsize),
            axis.title.y=element_text(size=axtsize),
            panel.grid=bgr)
    
  }
  
  if(dir=="greater"){
    
    basic <- ggplot(set, aes(x=grpf, y=mu)) +
      scale_x_discrete(labels=names) +
      xlab(xlabel) +
      ylab(ylabel) +
      geom_rect(aes(xmin=grp-0.5, xmax=grp+0.5, ymin=ldl, ymax=udl),
                alpha=0.5, fill=fillcol, linetype=0) +
      geom_segment(aes(x=grp-0.5, xend=grp+0.5, y=udl, yend=udl), size=lwidth, linetype=dlstyle) +
      geom_segment(aes(x=0.5, xend=max(grp)+0.5, y=gm, yend=gm), size=lwidth) +
      geom_segment(aes(x=grp, xend=grp, y=mu, yend=gm), size=lwidth) +
      geom_point(size=psize) +
      annotate("text", label="UDL", x=max(grp)+0.4, y=udl[max(grp)], size=4, vjust=-0.75) +
      ylim((min(mu)-(gm-min(mu))/5)[1],
           (max(max(mu), max(udl))+(max(max(mu), max(udl))-gm)/5)[1]) +
      back +
      theme(axis.text.x=element_text(size=axlsize),
            axis.text.y=element_text(size=axlsize),
            axis.title.x=element_text(size=axtsize),
            axis.title.y=element_text(size=axtsize),
            panel.grid=bgr)
    
  }
  
  if(dir=="less"){
    
    basic <- ggplot(set, aes(x=grpf, y=mu)) +
      scale_x_discrete(labels=names) +
      xlab(xlabel) +
      ylab(ylabel) +
      geom_rect(aes(xmin=grp-0.5, xmax=grp+0.5, ymin=ldl, ymax=udl),
                alpha=0.5, fill=fillcol, linetype=0) +
      geom_segment(aes(x=grp-0.5, xend=grp+0.5, y=ldl, yend=ldl), size=lwidth, linetype=dlstyle) +
      geom_segment(aes(x=0.5, xend=max(grp)+0.5, y=gm, yend=gm), size=lwidth) +
      geom_segment(aes(x=grp, xend=grp, y=mu, yend=gm), size=lwidth) +
      geom_point(size=psize) +
      annotate("text", label="LDL", x=max(grp)+0.4, y=ldl[max(grp)], size=4, vjust=1.5) +
      ylim((min(min(mu), min(ldl))-(gm-min(min(mu), min(ldl)))/5)[1],
           (max(mu)+(max(mu)-gm)/5)[1]) +
      back +
      theme(axis.text.x=element_text(size=axlsize),
            axis.text.y=element_text(size=axlsize),
            axis.title.x=element_text(size=axtsize),
            axis.title.y=element_text(size=axtsize),
            panel.grid=bgr)
    
  }
  
  if(dir=="greater"){
    npart <- geom_text(aes(y=min(mu), label=paste("n=", n, sep="")), vjust=3, size=npsize)
  }else{
    npart <- geom_text(aes(y=min(min(mu), min(ldl)), label=paste("n=", n, sep="")), vjust=3, size=npsize)
  }
  
  if(dir=="less"){
    ppart <- geom_text(aes(y=max(mu), label=p), vjust=-2, size=npsize)
  }else{
    ppart <- geom_text(aes(y=max(max(mu), max(udl)), label=p), vjust=-2, size=npsize)
  }
  
  if(printn==TRUE){
    if(printp==TRUE){
      chart <- basic + npart + ppart
    }else{
      chart <- basic + npart
    }
  }else{ 
    if(printp==TRUE){
      chart <- basic + ppart
    }else{
      chart <- basic
    }
  }
  
  print(chart)
  
}

R/ANOMintern.R
ANOMintern <- function(mu, n=NULL, gm=NULL, lo, up, names=NULL, alternative="two.sided",
                       xlabel="Group", ylabel="Endpoint", printn=TRUE, p=NULL, bg="white",
                       bgrid=TRUE, axlsize=18, axtsize=25, npsize=5, psize=5, lwidth=1,
                       dlstyle="dashed", fillcol="darkgray", whichone){
  
  whichone <- match.arg(whichone, choices=c("glm", "ratio"))
  
  bg <- match.arg(bg, choices=c("gray", "grey", "white"))
  
  if(bg=="white"){
    back <- theme_bw()
  }else{
    back <- theme_gray()
  }
  
  if(bgrid==TRUE){
    bgr <- element_line()
  }else{
    bgr <- element_blank()
  }
  
  if(whichone=="ratio"){
    
    if(is.null(gm)){
      gm <- weighted.mean(mu, n)
    }
    
    if(!(is.null(n)) & !(is.null(gm))){
      check <- abs(weighted.mean(mu, n) - gm)
      if(check > 0.01){
        stop("The 'gm' value you inserted is not the grand mean\n
             computed from your 'mu' and 'n' values!")
      }
      }
    
    }
  
  grp <- 1:length(mu)
  grpf <- as.factor(grp)
  
  if(!(is.null(names))){
    names <- as.factor(names)
  }else{
    names <- grpf
  }
  
  if(is.null(n)){
    printn <- FALSE
  }
  
  if(!(is.null(p))){
    printp <- TRUE
    p <- paste("p=", round(p, 3), sep="")
    p[p=="p=0"] <- "p<0.001"
  }else{
    printp <- FALSE
  }
  
  dir <- match.arg(alternative, choices=c("two.sided", "greater", "less"))
  
  if(dir=="two.sided"){
    if(all(lo==-Inf) | all(lo==-1) | all(lo==0) | all(up==1) | all(up==Inf)){
      warning("Is your alternative really 'two.sided'? Doesn't seem so!")
    }
  }
  
  if(whichone=="ratio"){
    ldl <- gm - abs(mu - up)
    udl <- gm + abs(mu - lo)    
  }else{
    ldl <- gm - abs((mu - gm) - up)
    udl <- gm + abs((mu - gm) - lo)
  }
  
  if(printp==TRUE){
    if(printn==TRUE){
      set <- data.frame(mu, gm, lo, up, ldl, udl, grp, grpf, names, n, p)
    }else{
      set <- data.frame(mu, gm, lo, up, ldl, udl, grp, grpf, names, p)
    }
  }else{
    if(printn==TRUE){
      set <- data.frame(mu, gm, lo, up, ldl, udl, grp, grpf, names, n)
    }else{
      set <- data.frame(mu, gm, lo, up, ldl, udl, grp, grpf, names)
    }
  }
  
  if(dir=="two.sided"){
    
    basic <- ggplot(set, aes(x=grpf, y=mu)) +
      scale_x_discrete(labels=names) +
      xlab(xlabel) +
      ylab(ylabel) +
      geom_rect(aes(xmin=grp-0.5, xmax=grp+0.5, ymin=ldl, ymax=udl),
                alpha=0.5, fill=fillcol, linetype=0) +
      geom_segment(aes(x=grp-0.5, xend=grp+0.5, y=udl, yend=udl), size=lwidth, linetype=dlstyle) +
      geom_segment(aes(x=grp-0.5, xend=grp+0.5, y=ldl, yend=ldl), size=lwidth, linetype=dlstyle) +
      geom_segment(aes(x=0.5, xend=max(grp)+0.5, y=gm, yend=gm), size=lwidth) +
      geom_segment(aes(x=grp, xend=grp, y=mu, yend=gm), size=lwidth) +
      geom_point(size=psize) +
      annotate("text", label="LDL", x=max(grp)+0.4, y=ldl[max(grp)], size=4, vjust=1.5) +
      annotate("text", label="UDL", x=max(grp)+0.4, y=udl[max(grp)], size=4, vjust=-0.75) +
      ylim(min(min(mu), min(ldl))-(gm-min(min(mu), min(ldl)))/5,
           max(max(mu), max(udl))+(max(max(mu), max(udl))-gm)/5) +
      back +
      theme(axis.text.x=element_text(size=axlsize),
            axis.text.y=element_text(size=axlsize),
            axis.title.x=element_text(size=axtsize),
            axis.title.y=element_text(size=axtsize),
            panel.grid=bgr)
    
  }
  
  if(dir=="greater"){
    
    basic <- ggplot(set, aes(x=grpf, y=mu)) +
      scale_x_discrete(labels=names) +
      xlab(xlabel) +
      ylab(ylabel) +
      geom_rect(aes(xmin=grp-0.5, xmax=grp+0.5, ymin=ldl, ymax=udl),
                alpha=0.5, fill=fillcol, linetype=0) +
      geom_segment(aes(x=grp-0.5, xend=grp+0.5, y=udl, yend=udl), size=lwidth, linetype=dlstyle) +
      geom_segment(aes(x=0.5, xend=max(grp)+0.5, y=gm, yend=gm), size=lwidth) +
      geom_segment(aes(x=grp, xend=grp, y=mu, yend=gm), size=lwidth) +
      geom_point(size=psize) +
      annotate("text", label="UDL", x=max(grp)+0.4, y=udl[max(grp)], size=4, vjust=-0.75) +
      ylim((min(mu)-(gm-min(mu))/5)[1],
           (max(max(mu), max(udl))+(max(max(mu), max(udl))-gm)/5)[1]) +
      back +
      theme(axis.text.x=element_text(size=axlsize),
            axis.text.y=element_text(size=axlsize),
            axis.title.x=element_text(size=axtsize),
            axis.title.y=element_text(size=axtsize),
            panel.grid=bgr)
    
  }
  
  if(dir=="less"){
    
    basic <- ggplot(set, aes(x=grpf, y=mu)) +
      scale_x_discrete(labels=names) +
      xlab(xlabel) +
      ylab(ylabel) +
      geom_rect(aes(xmin=grp-0.5, xmax=grp+0.5, ymin=ldl, ymax=udl),
                alpha=0.5, fill=fillcol, linetype=0) +
      geom_segment(aes(x=grp-0.5, xend=grp+0.5, y=ldl, yend=ldl), size=lwidth, linetype=dlstyle) +
      geom_segment(aes(x=0.5, xend=max(grp)+0.5, y=gm, yend=gm), size=lwidth) +
      geom_segment(aes(x=grp, xend=grp, y=mu, yend=gm), size=lwidth) +
      geom_point(size=psize) +
      annotate("text", label="LDL", x=max(grp)+0.4, y=ldl[max(grp)], size=4, vjust=1.5) +
      ylim((min(min(mu), min(ldl))-(gm-min(min(mu), min(ldl)))/5)[1],
           (max(mu)+(max(mu)-gm)/5)[1]) +
      back +
      theme(axis.text.x=element_text(size=axlsize),
            axis.text.y=element_text(size=axlsize),
            axis.title.x=element_text(size=axtsize),
            axis.title.y=element_text(size=axtsize),
            panel.grid=bgr)
    
  }
  
  if(dir=="greater"){
    npart <- geom_text(aes(y=min(mu), label=paste("n=", n, sep="")), vjust=3, size=npsize)
  }else{
    npart <- geom_text(aes(y=min(min(mu), min(ldl)), label=paste("n=", n, sep="")), vjust=3, size=npsize)
  }
  
  if(dir=="less"){
    ppart <- geom_text(aes(y=max(mu), label=p), vjust=-2, size=npsize)
  }else{
    ppart <- geom_text(aes(y=max(max(mu), max(udl)), label=p), vjust=-2, size=npsize)
  }
  
  if(printn==TRUE){
    if(printp==TRUE){
      chart <- basic + npart + ppart
    }else{
      chart <- basic + npart
    }
  }else{ 
    if(printp==TRUE){
      chart <- basic + ppart
    }else{
      chart <- basic
    }
  }
  
  print(chart)
  
}